clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -505,68 +536,110 @@ public void fireOnAssignment(Set<TopicPartition> assignment) {
+            this.state = FetchState.INITIALIZING;
             this.position = null;
             this.highWatermark = null;
             this.logStartOffset = null;
             this.lastStableOffset = null;
             this.resetStrategy = null;
-            this.nextAllowedRetryTimeMs = null;
+            this.nextRetryTimeMs = null;
         }
 
         private void reset(OffsetResetStrategy strategy) {
+            this.state = FetchState.AWAIT_RESET;
             this.resetStrategy = strategy;
             this.position = null;
-            this.nextAllowedRetryTimeMs = null;
+            this.nextRetryTimeMs = null;
+        }
+
+        private boolean maybeValidatePosition(Metadata.LeaderAndEpoch currentLeader) {
+            if (!hasPosition())
+                throw new IllegalStateException("Cannot validate offset while partition is in state " + state);
+
+            this.position = new FetchPosition(position.offset, position.lastFetchEpoch, currentLeader);
+            // If we have no epoch information for the current position, then we can skip validation.
+            if (position.lastFetchEpoch.isPresent()) {
+                this.state = FetchState.AWAIT_VALIDATION;
+                return true;
+            } else {
+                this.state = FetchState.FETCHING;
+                return false;
+            }
+        }
+
+        /**
+         * Clear the awaiting validation state and enter fetching.
+         */
+        private void validate() {
+            if (hasPosition()) {
+                this.state = FetchState.FETCHING;
+                this.nextRetryTimeMs = null;
+            }
         }
 
-        private boolean isResetAllowed(long nowMs) {
-            return nextAllowedRetryTimeMs == null || nowMs >= nextAllowedRetryTimeMs;
+        private boolean awaitingValidation() {
+            return state == FetchState.AWAIT_VALIDATION;
+        }
+
+        private boolean awaitingRetryBackoff(long nowMs) {
+            return nextRetryTimeMs != null && nowMs < nextRetryTimeMs;
         }
 
         private boolean awaitingReset() {
-            return resetStrategy != null;
+            return state == FetchState.AWAIT_RESET;
         }
 
-        private void setResetPending(long nextAllowedRetryTimeMs) {
-            this.nextAllowedRetryTimeMs = nextAllowedRetryTimeMs;
+        private void setNextAllowedRetry(long nextAllowedRetryTimeMs) {
+            this.nextRetryTimeMs = nextAllowedRetryTimeMs;
         }
 
         private void resetFailed(long nextAllowedRetryTimeMs) {
-            this.nextAllowedRetryTimeMs = nextAllowedRetryTimeMs;
+            this.nextRetryTimeMs = nextAllowedRetryTimeMs;
         }
 
         private boolean hasValidPosition() {
-            return position != null;
+            return state == FetchState.FETCHING;
         }
 
-        private boolean isMissingPosition() {
-            return !hasValidPosition() && !awaitingReset();
+        private boolean hasPosition() {
```

Comment:
It's a little confusing to say we have a position while it is still being reset. Maybe it could be `isInitializing` or something like that?